vpr/src/pack/prepack.cpp

```diff
@@ -742,65 +817,75 @@ t_pack_molecule *alloc_and_load_pack_molecules(
-                    cur_was_last_inserted = (last_valid_iter->second == cur_molecule);
-                }
-                if(range_empty || !cur_was_last_inserted) {
-                    /* molecule did not cover current atom (possibly because molecule created is
-                     * part of a long chain that extends past multiple logic blocks), try again */
-                    --blk_iter;
-                }
-            }
+		/* Find forced pack patterns
+		 * Simplifying assumptions: Each atom can map to at most one molecule,
+		 */
+		for (int pattern = 0; pattern < num_packing_patterns; pattern++) {
+			feasible_patterns[pattern] = check_atom_feasible_for_pattern(blk_id, &list_of_pack_patterns[pattern]);
+		}
+		best_pattern = find_best_pattern(blk_id, list_of_pack_patterns, feasible_patterns, num_packing_patterns, atom_molecules);
+		if(best_pattern >= 0) {
+			bool cur_was_last_inserted = true;
+			int num_packed_atoms = 0;
+			do {
+				cur_molecule = try_create_molecule(list_of_pack_patterns, atom_molecules, best_pattern, blk_id, &num_packed_atoms);
+				if(cur_molecule != nullptr) {
+					list_of_pack_patterns[best_pattern].usage++;
+					cur_molecule->next = list_of_molecules_head;
+					/* In the event of multiple molecules with the same atom block pattern,
+					 * bias to use the molecule with less costly physical resources first */
+					/* TODO: Need to normalize magical number 100 */
+					cur_molecule->base_gain = cur_molecule->num_blocks - (cur_molecule->pack_pattern->base_cost / 100);
+					list_of_molecules_head = cur_molecule;
+
+					//Note: atom_molecules is an (ordered) multimap so the last molecule
+					//      inserted for a given blk_id will be the last valid element
+					//      in the equal_range
+					auto rng = atom_molecules.equal_range(blk_id); //The range of molecules matching this block
+					bool range_empty = (rng.first == rng.second);
+					if (!range_empty) {
+						auto last_valid_iter = --rng.second; //Iterator to last element (only valid if range is not empty)
+						cur_was_last_inserted = (last_valid_iter->second == cur_molecule);
+					}
+					if(range_empty || !cur_was_last_inserted) {
+					    /* molecule did not cover current atom (possibly because molecule created is
+					     * part of a long chain that extends past multiple logic blocks), try again */
+					    --blk_iter;
+					} else {
+						/* try_pack_molecule function can pack more than one atom */
+						blk_iter += num_packed_atoms;
+						blk_id = *(blk_iter);
+						/* check if we packed last atom of a chain */
+						if(blk_iter == blocks.end()) {
+							/* no more atoms to pack */
+							cur_was_last_inserted = true;
+						} else {
+							/* check if the next block is a part of the chain we're currently packing */
+							if((check_blocks_in_chain(*(blk_iter-1), *(blk_iter), &list_of_pack_patterns[best_pattern]))) {
+								cur_was_last_inserted = false;
+							}
```

Comment:
Move this into a function?

Reply:
the loop manipulates blk iterator of the main loop. Moving it to a funcion may complicate the code

Reply:
Ya, I see that.  It is disappointing :(